mitodl/mit-open
# Add a channel for every topic, department, offeror

### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/3978

### Description (What does it do?)
- Adds a `ChannelPlugin` class with functions to create channels of the appropriate type  (topic, department, or offeror).
- Calls the plugin topic function when a new topic is created in `learning_resources.etl.load_topics()`
- Calls the plugin offeror & department functions as part of the `updated_offered_by` and `update_departments` management commands respectively.
- Adds a new `backpopulate_resource_channels` that will create channels of the appropriate type for all the above, or just the specified type via options.



### How can this be tested?
- In django admin, delete the "Linear Algebra" `LearningResourceTopic` if it exists, then run `./manage.py backpopulate_ocw_data --course-name 18-700-linear-algebra-fall-2013
- Once done, the topic should exist again, and there should be a new `FieldChannel` with `channel_type=topic`, `name=linear-algebra`, `title=Linear Algebra`, and an associated `ChannelTopicDetail` with `topic=<id of the topic>`
- Run `./manage.py update_departments`.   Once complete, there should be a  `FieldChannel` with `channel_type=department` for each `LearningResourceDepartment` with relevant `name`, `title` and `ChannelDepartmentDetail` for each.
- Run `./manage.py update_offered_by`.   Once complete, there should be a  `FieldChannel` with `channel_type=offeror` for each `LearningResourceOfferor` with relevant `name`, `title` and `ChannelOfefrorDetail` for each.
- Delete all the new field channels via django admin.
- Run `./manage.py backpopulate_resource_channels`.  When done, there should be a `FieldChannel` for every department, offeror, and topic.

Try these urls, they should work and show appropriately filtered search results:

- http://localhost:8063/c/department/civil-and-environmental-engineering
- http://localhost:8063/c/topic/systems-engineering
- http://localhost:8063/c/offeror/ocw
